### PR TITLE
Jenkins longest job stage analysis script

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
Signed-off-by: Andrew Leonard <anleonar@redhat.com>

### CODE REVIEW

File: tools/stage_time_analysis.groovy [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+          getJobs(stats_file, find_stage, "${folder}/job/${job.name}", job.jobs)
+        }
+      } else {
+        def jobName="${folder}/job/${job.name}"
+        if (folder.equals("")) {
+          jobName="job/${job.name}"
+        }
+        println("${jobName}")
+
+        def workflow = null
```
Comment (karianna) [inline tools/stage_time_analysis.groovy:32]: extract this section to its own function?

File: tools/stage_time_analysis.groovy [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        try {
+          workflow = sh(returnStdout: true, script: "wget -q -O - ${params.JENKINS_URL}/${jobName}/lastSuccessfulBuild/wfapi/describe")
+        } catch(Exception e) {
+          workflow = null
+        }
+        if (workflow) {
+          def json = new JsonSlurper().parseText(workflow)
+          json.stages.each { stage ->
+            if (stage.name.equalsIgnoreCase(find_stage)) {
+              def duration = stage.durationMillis/60000
```
Comment (karianna) [inline tools/stage_time_analysis.groovy:42]: extract 60000 to a named const

File: tools/stage_time_analysis.groovy [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  def jenkinsUrl = "${params.JENKINS_URL}"
+  def find_stage = "${params.STAGE}"
+  def stats_file = "stage.stats"
+
+  stage("getStageStats") {
+    sh("rm -f ${stats_file}")
+
+    try {
+      // Get all jobs list
+      def jobs = sh(returnStdout: true, script: "wget -q -O - ${jenkinsUrl}/api/json?tree=jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name]]]]]]]]]]")
```
Comment (karianna) [inline tools/stage_time_analysis.groovy:65]: This seems brittle (the chain of name,job, name, job...

**Review by karianna — CHANGES_REQUESTED** [review]

File: tools/stage_time_analysis.groovy (hunk 3 above)
Comment (andrew-m-leonard) [reply tools/stage_time_analysis.groovy:65]: Yeah, the API is somewhat limited to return sub-folders, i've allowed a depth of 10 folders, which is probably reasonable.

File: tools/stage_time_analysis.groovy [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import groovy.json.JsonSlurper
+
+  def getJobs(String stats_file, String find_stage, String folder, Object list_jobs) {
```
[the remark this answers is not in the document]
Comment (andrew-m-leonard) [reply tools/stage_time_analysis.groovy]: changed

File: tools/stage_time_analysis.groovy [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    list_jobs.each { job ->
+      if (job.jobs && job.jobs.size() > 0) {
+        if (folder.equals("")) {
```
[the remark this answers is not in the document]
Comment (andrew-m-leonard) [reply tools/stage_time_analysis.groovy]: possibly, updated

File: tools/stage_time_analysis.groovy (hunk 1 above)
Comment (andrew-m-leonard) [reply tools/stage_time_analysis.groovy:21]: extracted to a fn

File: tools/stage_time_analysis.groovy (hunk 2 above)
Comment (andrew-m-leonard) [reply tools/stage_time_analysis.groovy]: changed

File: tools/stage_time_analysis.groovy (hunk 3 above)
Comment (karianna) [reply tools/stage_time_analysis.groovy:69]: Cool, can you enhance the comment explain thy we can't have nice things? :-)

**Comment by andrew-m-leonard** [comment]
Test run: https://ci.adoptopenjdk.net/view/work-in-progress/job/andrew-stage-analysis/26/console

**Review by sxa — APPROVED** [review]
On the basis that it runs in your job and appears to produce valid output, I approve

**Review by karianna — APPROVED** [review]
LGTM

**Pull request merged**